### PR TITLE
RAS: add new case for RAS feature support

### PR DESCRIPTION
**Pull request by xiaodwan**

### CODE REVIEW

**Comment by xiaodwan** [comment]
depends on https://github.com/avocado-framework/avocado-vt/pull/3960

**Comment by Yingshun** [comment]
@xiaodwan  plz upload test results, thx!

File: libvirt/tests/cfg/features/ras.cfg [hunk 1]
```diff
@@ -0,0 +1,12 @@
+- features.ras:
```
Comment (Yingshun) [inline libvirt/tests/cfg/features/ras.cfg:1]: Should this test belong to cpu? if yes, can you plz move it to cpu folder?

File: libvirt/tests/src/features/ras.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import re
+import logging as log
+
+from virttest.libvirt_xml.vm_xml import VMXML
+from virttest import virsh
+from virttest import libvirt_version
+
+# Using as lower capital is not the best way to do, but this is just a
+# workaround to avoid changing the entire file.
+logging = log.getLogger('avocado.' + __name__)
```
Comment (Yingshun) [inline libvirt/tests/src/features/ras.py:10]: As you can see from the comments, we named it as 'logging' as a workaround, so for new files, we'd better have another one, like 'LOG'.

File: libvirt/tests/src/features/ras.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+def run(test, params, env):
+    """
+    Test the ras feature
+    1. Enable 'ras=on/off' in a guest
+    2. check the xml and qemu cmd line
+    """
+    # Ras feature supported since 10.4.0.
+    if not libvirt_version.version_compare(10, 4, 0):
+        test.cancel("Ras feature is not supported "
+                    "on current version.")
```
Comment (Yingshun) [inline libvirt/tests/src/features/ras.py:22]: I would suggest to update it to 'libvirt_version.is_libvirt_feature_supported(params)' and add 'func_supported_since_libvirt_ver = (10, 4, 0)' in cfg.

File: libvirt/tests/src/features/ras.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    def check_dumpxml():
+        """
+        Check whether the added devices are shown in the guest xml
+        """
+        pattern = "<ras state=\"%s\" />" % ras_state
+        # Check ras state
+        vm_xml = VMXML.new_from_dumpxml(vm_name)
+        if pattern not in str(vm_xml):
+            test.fail("Can not find %s "
+                      "in the guest xml file." % pattern)
```
Comment (Yingshun) [inline libvirt/tests/src/features/ras.py:35]: The recommended way to check the guest's xml is libvirt_vmxml.check_guest_xml_by_xpaths, can you plz check if you can reuse it?

File: libvirt/tests/cfg/features/ras.cfg (hunk 1 above)
Comment (dzhengfy) [reply libvirt/tests/cfg/features/ras.cfg:1]: The attributes is in <features>, so it does not belong to cpu.
In addition, there is vm_features.py in cpu folder. I think we should move it to the features folder. What do you think? @Yingshun

File: libvirt/tests/src/features/ras.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    def check_qemu_cmd_line():
```
Comment (dzhengfy) [inline libvirt/tests/src/features/ras.py:37]: Please reuse` libvirt.check_qemu_cmd_line()`

File: libvirt/tests/src/features/ras.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        """
+        Check whether the ras feature is shown in the qemu cmd line
+        """
+        if not vm.get_pid():
+            test.fail('VM pid file missing.')
+        with open('/proc/%s/cmdline' % vm.get_pid()) as cmdline_file:
```
Comment (dzhengfy) [inline libvirt/tests/src/features/ras.py:43]: Do you need to check qemu command line or vm kernel command line? please double check.

File: libvirt/tests/src/features/ras.py (hunk 5 above)
Comment (xiaodwan) [reply libvirt/tests/src/features/ras.py:37]: ok

File: libvirt/tests/src/features/ras.py (hunk 6 above)
Comment (xiaodwan) [reply libvirt/tests/src/features/ras.py:43]: libvirt.check_qemu_cmd_line() will do.

File: libvirt/tests/src/features/ras.py (hunk 4 above)
Comment (xiaodwan) [reply libvirt/tests/src/features/ras.py:35]: ack

File: libvirt/tests/src/features/ras.py (hunk 3 above)
Comment (xiaodwan) [reply libvirt/tests/src/features/ras.py:22]: thanks

File: libvirt/tests/src/features/ras.py (hunk 2 above)
Comment (xiaodwan) [reply libvirt/tests/src/features/ras.py:10]: thanks

**Comment by xiaodwan** [comment]
```
JOB ID     : a5954ec4c75aadf4f5d9ee9f470e598d549851dd
JOB LOG    : /var/log/avocado/job-results/job-2024-09-03T02.11-a5954ec/job.log
 (1/2) type_specific.io-github-autotest-libvirt.features.ras.positive_test.on: STARTED
 (1/2) type_specific.io-github-autotest-libvirt.features.ras.positive_test.on: PASS (7.89 s)
 (2/2) type_specific.io-github-autotest-libvirt.features.ras.positive_test.on: STARTED
 (2/2) type_specific.io-github-autotest-libvirt.features.ras.positive_test.on: PASS (8.73 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/log/avocado/job-results/job-2024-09-03T02.11-a5954ec/results.html
JOB TIME   : 18.33 s

JOB ID     : 7642ac7b241fed17949d7b7e8d5ca7e1bdc0f7fc
JOB LOG    : /var/log/avocado/job-results/job-2024-09-03T02.11-7642ac7/job.log
 (1/2) type_specific.io-github-autotest-libvirt.features.ras.positive_test.off: STARTED
 (1/2) type_specific.io-github-autotest-libvirt.features.ras.positive_test.off: PASS (7.93 s)
 (2/2) type_specific.io-github-autotest-libvirt.features.ras.positive_test.off: STARTED
 (2/2) type_specific.io-github-autotest-libvirt.features.ras.positive_test.off: PASS (8.75 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/log/avocado/job-results/job-2024-09-03T02.11-7642ac7/results.html
```

File: libvirt/tests/src/features/ras.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
+import logging
+
+from virttest import virsh
+from virttest import libvirt_version
+from virttest.libvirt_xml.vm_xml import VMXML
+from virttest.utils_libvirt import libvirt_vmxml
+from virttest.utils_test import libvirt
+
+# Using as lower capital is not the best way to do, but this is just a
```
Comment (Yingshun) [inline libvirt/tests/src/features/ras.py:9]: Please remove the unnecessary comments, thx!

File: libvirt/tests/src/features/ras.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    LOG.debug("vm xml is %s", vm_xml_backup)
+
+    if vm.is_alive():
+        vm.destroy()
+
+    try:
+        features_xml = vm_xml.features
+        if features_xml.has_feature('ras'):
+            features_xml.remove_feature('ras')
+        features_xml.ras = "%s" % ras_state
```
Comment (Yingshun) [inline libvirt/tests/src/features/ras.py:55]: Can it be 'features_xml.ras = ras_state'?

**Review by Yingshun — COMMENTED** [review]
Others LGTM

File: libvirt/tests/src/features/ras.py (hunk 7 above)
Comment (xiaodwan) [reply libvirt/tests/src/features/ras.py]: Done

File: libvirt/tests/src/features/ras.py (hunk 8 above)
Comment (xiaodwan) [reply libvirt/tests/src/features/ras.py]: Done

**Review by Yingshun — APPROVED** [review]

**Review by dzhengfy — APPROVED** [review]
LGTM

**Pull request merged**